### PR TITLE
SURF-623: Modify Grid

### PR DESCRIPTION
**Pull request by CITguy**
JIRA: https://jira.rax.io/browse/SURF-623

### LGTM's
- [x] Dev LGTM
- [x] Zoom LGTM

#### TL;DR

* `.hx{modifier}[-{size}]-{N}` renamed to `.hx{modifier}-{N}[-{size}]`
  * optional piece moved from _middle_ to _end_ of class name
* `.hxCol` split into `.hxCol` and `.hxSpan-{N}[-{size}]`
  * separates _content layout_ from _container sizing_
* `.hxCol` no longer applies a columnar flexbox layout. 
  * The majority of the time this is unneeded.
  * It changed some built-in browser behavior (margin collapsing; caused more space than necessary to be added to the document).
* added `.hxRow.hxGutterless` styling
* clarified documentation

#### BEFORE
* Container and width modifier classes were tightly coupled (can't use width modifiers without turning an element into a Column container).
* Documentation unclear when to use Row and Column containers.
* Modifier classes had their optional parts before the required parts
  * `.hxModifier[-{size}]-{N}`

grid-before

#### AFTER
* Clearer documentation
* Modifier classes now have their optional parts at the end of the class name.
  * `.hxModifier-{N}[-{size}]`
* Gutterless rows now possible with `.hxRow.hxGutterless`

grid-after

### CODE REVIEW

File: bin/start.js [hunk 1]
```diff
@@ -35,16 +35,6 @@ browserSync.init({
...
-
-                /**
-                 * Always ensure that this task runs last, as it needs to wait for all
-                 * synchronous build events to complete so that browserSync can finish
-                 * and reload the page (and therefore, allow selenium to connnect).
-                 */
-                const tests = exec('yarn run test:regression -u', { cwd: CONFIG.testDir });
-                // this *is* asynchronous, see comment above
-                tests.stdout.pipe(process.stdout);
-                tests.stderr.pipe(process.stderr);
```
Comment (CITguy) [inline bin/start.js]: @Droogans: This was removed because the process almost locked up my machine when moving files around while the dev server was running.

File: source/components/grid/grid.less [hunk 2]
```diff
@@ -79,21 +95,25 @@
     margin-top: @grid-gutter-size;
   }
 
-  .hxCol,
-  [class*="hxCol-"] {
+  & > .hxCol {
```
Comment (CITguy) [inline source/components/grid/grid.less]: Only apply gutters to immediate children.

**Comment by mcortez1497** [comment]
Could we change the `gutterless` attribute to be a class name instead?

File: source/components/grid/index.html [hunk 3]
```diff
@@ -5,363 +5,263 @@
...
+    The Grid CSS module provides utility classes based on flexbox to help
+    with the positioning and layout of content within containers.
   </p>
+</section>
+
+<section><!-- rows -->
+  <h2 class="hxSectionTitle" id="rows">Rows</h2>
+  <ul>
+    <li>
+      Rows are containers that allow you to divid their
```
Comment (nimbinatus) [inline source/components/grid/index.html]: s/`divid`/`divide`

File: source/components/grid/index.html [hunk 4]
```diff
@@ -415,232 +307,283 @@ <h3 class="hxSubSectionTitle" id="responsive">Responsive Layout</h3>
...
-    <code>padding</code>, or geometry of these elements will affect the
-    layout.
-  </p>
-
   <ul>
-    <li>
-      Wrap your content in a <code>&lt;div&gt;</code> and style
-      using a custom CSS class.
-    </li>
+    <li>Rows have margins set to align the outer Columsn with surrounding content.</li>
```
Comment (mcortez1497) [inline source/components/grid/index.html]: *Columns

File: source/components/grid/index.html [hunk 5]
```diff
@@ -5,363 +5,263 @@
...
-      See <a href="#wrapping">Column Wrapping</a> for more details.
-    </div>
-  </div>
-  <br />
-
+<section><!-- column width -->
+  <h2 class="hxSectionTitle" id="column-width">Column Width</h2>
   <p>
-    Explicit columns will take up a defined percentage of the row.
+    Column width modifieres will set the width of a Column container to a
```
Comment (nimbinatus) [inline source/components/grid/index.html]: s/`modifieres`/`modifiers`

Comment (mcortez1497) [inline source/components/grid/index.html]: *modifiers

File: source/components/grid/index.html (hunk 3 above)
Comment (mcortez1497) [inline source/components/grid/index.html]: *divide

File: source/components/grid/index.html [hunk 6]
```diff
@@ -382,28 +282,20 @@ <h3 class="hxSubSectionTitle" id="reordering">Reordering Columns</h3>
...
-    If you mix responsive clases, larger breakpoints will trump smaller ones.
-  </p>
-
+<section><!-- responsive options -->
+  <h2 class="hxSectionTitle" id="responsive-options">Responsive Options</h2>
+  <ul>
+    <li>All Column modifiers support adding an optional breakpoint</li>
+    <li>
+      Omitting the breakpoint will default to <code>xs</code>
+      (applies to all breakpoints)
```
Comment (nimbinatus) [inline source/components/grid/index.html]: Period needed at the end of this line.

File: source/components/grid/index.html [hunk 7]
```diff
@@ -382,28 +282,20 @@ <h3 class="hxSubSectionTitle" id="reordering">Reordering Columns</h3>
...
+    </li>
+    <li>
+      Each respoinsive class corresponds to the <em>minimum</em> breakpoint
+      required to achieve the desired appearance.
+    </li>
+    <li>Larger breakpoints override smaller ones</li>
```
Comment (nimbinatus) [inline source/components/grid/index.html]: Period needed at the end of this line.

File: source/components/grid/index.html [hunk 8]
```diff
@@ -415,232 +307,283 @@ <h3 class="hxSubSectionTitle" id="responsive">Responsive Layout</h3>
...
-  <h3 class="hxSubSectionTitle" id="wrapping">Column Wrapping</h3>
+<section><!-- column wrapping -->
+  <h2 class="hxSectionTitle" id="column-wrapping">Column Wrapping</h2>
+  <div class="hxAlert">
+    <div class="hxAlert__icon">
+      <hx-icon type="info-circle"></hx-icon>
+    </div>
+    <div class="hxAlert__text">
+      <span class="hxAlert__status">NOTE</span>
+      Columns wrap if their total column count exceeds 12
```
Comment (nimbinatus) [inline source/components/grid/index.html]: Period needed at the end of this line.

Comment (nimbinatus) [reply source/components/grid/index.html]: Also, suggest something like this:
```html
Columns wrap if the total column width count within the row exceeds 12.
```

File: source/components/grid/index.html (hunk 4 above)
Comment (nimbinatus) [inline source/components/grid/index.html]: s/`Columsn`/`Columns` (also, why are `Columns` and `Rows` capitalized globally?

Comment (CITguy) [reply source/components/grid/index.html]: They are capitalized to promote distinction as a type of container, rather than the more generic concept of a row/column.

Comment (CITguy) [reply source/components/grid/index.html]: Now that you mention this, I'm going to go ahead and modify the capitalization.

**Comment by mcortez1497** [comment]
Dev LGTM

File: bin/start.js (hunk 1 above)
Comment (Droogans) [reply bin/start.js]: Spent some time looking closer at this and I'm pretty sure this isn't going to fix the problem completely.

I've noticed that each file change event triggers these actions, meaning moving 10 files around will fire off 10 simultaneous build tasks. I'll look at limiting it globally with some sort of debounce or throttle.

Comment (Droogans) [reply bin/start.js]: https://browsersync.io/docs/options#option-reloadDebounce
https://browsersync.io/docs/options#option-reloadThrottle
